### PR TITLE
Enable one-click booking to NOMIS

### PR DESCRIPTION
**Pull request by alan**

### CODE REVIEW

File: spec/decorators/visit_decorator_spec.rb [hunk 1]
```diff
@@ -35,4 +35,76 @@
...
+
+    before do
+    end
+
+    describe 'when the book to nomis feature is disabled' do
+      before do
+        switch_off :nomis_staff_book_to_nomis_enabled
+      end
+
+      it { expect(subject.book_to_nomis_opt_in?).to eq(false) }
```
Comment (StupidCodeFactory) [inline spec/decorators/visit_decorator_spec.rb]: 1) Although this is correct the canonical way to check a boolean in RSpec is:

```ruby
it { expect(subject.book_to_nomis_opt_in?).to be false }
```
2) Also when testing predicate methods, RSpec provide a very nice and readable syntactic sugar API:

```ruby
it { is_expected.not_to be_book_to_nomis_opt_in }
```

File: spec/services/nomis/api_spec.rb [hunk 2]
```diff
@@ -185,4 +185,41 @@
...
+      }
+    end
+
+    let(:offender_id) { 1_057_307 }
+
+    subject { super().book_visit(offender_id: offender_id, params: params) }
+
+    context 'happy path', vcr: { cassette_name: 'book_visit_happy_path' } do
+      it 'returns the visit_id' do
+        expect(subject.visit_id).to eq(5_467)
```
Comment (StupidCodeFactory) [inline spec/services/nomis/api_spec.rb]: RSpec provides a nice helper to do this more elegantly:
```ruby
expect(subject.book_visit(offender_id: offender_id, params: params)).to have_attributes(visit_id: 5_467)
```

File: spec/services/nomis/api_spec.rb [hunk 3]
```diff
@@ -185,4 +185,41 @@
...
+      end
+
+      it 'instruments the outcome of the call' do
+        subject
+        expect(PVB::Instrumentation.custom_log_items[:book_to_nomis_success]).to eq(true)
```
Comment (StupidCodeFactory) [inline spec/services/nomis/api_spec.rb]: I think it is clearer as you directly see what method call is being tested with which arguments and also remove this odd call to subject on the first line of the test:

```ruby
expect {
  subject.book_visit(offender_id: offender_id, params: params)
}.to change { PVB::Instrumentation.custom_log_items[:book_to_nomis_success] }.from(nil).to(true)

File: spec/services/nomis/feature_spec.rb [hunk 4]
```diff
@@ -107,4 +107,32 @@
...
     end
   end
+
+  describe '.book_to_nomis_enabled?' do
+    context 'when the book to nomis flag is disabled' do
+      before do
+        switch_off :nomis_staff_book_to_nomis_enabled
+      end
+
+      it { expect(described_class.book_to_nomis_enabled?(anything)).to eq(false) }
```
Comment (StupidCodeFactory) [inline spec/services/nomis/feature_spec.rb]:
```ruby
 is_expected.not_to be_book_to_nomis_enabled anything
```

Comment (alan) [inline spec/services/nomis/feature_spec.rb]: I find the use of predicate methods with variables not very readable when they take parameters. I think is clearer like it is now.

File: spec/decorators/visit_decorator_spec.rb (hunk 1 above)
Comment (alan) [inline spec/decorators/visit_decorator_spec.rb]: The use of predicate methods has the downside that you lose the context of the exact method call. I think the tests end up being more maintainable by using less RSpec sugar methods, my preference would even to use mini_test!

File: app/views/prison/visits/process_visit.html.erb [hunk 5]
```diff
@@ -11,6 +11,7 @@
     <%= hidden_field_tag 'visit[visitor_ids][]', visitor.id  %>
   <% end %>
   <%= f.hidden_field :prison_id  %>
+  <%# f.hidden_field_tag 'prisoner[nomis_offender_id]', @visit.nomis_offender_id %>
```
Comment (StupidCodeFactory) [inline app/views/prison/visits/process_visit.html.erb]: Could this be removed?

Comment (chrishanes) [inline app/views/prison/visits/process_visit.html.erb]: @alan mentioned we needed the offender_id as a hidden field, but left this commented out.

File: .rspec [hunk 6]
```diff
@@ -1,3 +1,2 @@
 --color
 --require spec_helper
---backtrace
```
Comment (StupidCodeFactory) [inline .rspec]: Should we add ```.rspec``` to the ignore file? I defo need the backtrace when doing any rspec testing

Comment (alan) [inline .rspec]: Yes, we could do that.

In theory you should be able to add it to your global .rspec file - https://relishapp.com/rspec/rspec-core/v/2-0/docs/configuration/read-command-line-configuration-options-from-files

Out of curiosity, how come you find it useful all the time? I've found it useful very few times

File: app/services/booking_response.rb [hunk 7]
```diff
@@ -1,9 +1,44 @@
...
+  def self.succesful
+    new(nil)
+  end
+
+  def self.process_required
+    new(PROCESS_REQUIRED_ERROR)
+  end
+
+  def self.nomis_validation_error
+    new(NOMIS_VALIDATION_ERROR)
```
Comment (StupidCodeFactory) [inline app/services/booking_response.rb]: these to `NOMIS_VALIDATION_ERROR` and `NOMIS_API_ERROR` should really be in the Nomis API service class instead of being bundled with other unrelated error messages

File: app/services/nomis/booker.rb [hunk 8]
```diff
@@ -0,0 +1,56 @@
+module Nomis
+  class Booker
+    ALREADY_PROCESSED = 'Duplicate post'.freeze
+
+    def initialize(visit)
+      self.visit = visit
```
Comment (StupidCodeFactory) [inline app/services/nomis/booker.rb]: I'm this class doesn't really need to know about the the ActiveRecord Visit model, it would be much nice to pass in the `booking_params` and have the caller know about the know about the visit rather then have both know about the internals of the `Visit` ActiveRecord object

File: app/views/layouts/application.html.erb [hunk 9]
```diff
@@ -64,9 +64,9 @@
     <% end %>
 
     <% if alert.present? %>
-        <p class="notification">
-            <%= alert %>
-        </p>
+      <div class="error-summary" role="group" aria-labelledby="error-summary-heading" tabindex="-1">
```
Comment (StupidCodeFactory) [inline app/views/layouts/application.html.erb]: How does this relate to booking to nomis? Can you refactor this in a separate PR. Also I can see any unit test for this change.

File: app/views/prison/visits/process_visit.html.erb [hunk 10]
```diff
@@ -68,10 +72,25 @@
         <strong class="bold-small"><%= t('.rejection_warning') %></strong>
       </div>
     </div>
-    <%= f.submit t('.submit'), class: 'button button-primary' %>
-    <div class="push-top">
-      <%= link_to t('.email_preview'), prison_visit_email_preview_path(@visit), class: 'js-LinkPreview', target: '_blank' %>
-    </div>
+    <% if Nomis::Feature.book_to_nomis_enabled?(@visit.prison_name) %>
```
Comment (StupidCodeFactory) [inline app/views/prison/visits/process_visit.html.erb]: We should stop treating our views as second class citizens and allow them to become such a mess.
This entire block should be in a decorator and be unit tested

File: spec/controllers/prison/visits_controller_spec.rb [hunk 11]
```diff
@@ -71,6 +71,42 @@
 
         it { is_expected.to redirect_to(prison_inbox_path) }
       end
+
+      context 'when there is an api error when booking to nomis' do
```
Comment (StupidCodeFactory) [inline spec/controllers/prison/visits_controller_spec.rb]: there need to be an outer scope called "When booking to nomis" as it seems to be repeated in the next context block, that would share the setup of `let(:stadd_response)` and `let(:booking_response)`

File: spec/decorators/visit_decorator_spec.rb [hunk 12]
```diff
@@ -11,6 +11,33 @@
...
+
+    context 'when the lookup is enabled' do
+      let(:checker) { double(StaffNomisChecker) }
+      let(:offender) { double(Nomis::Offender, id: 1234) }
+
+      before do
+        switch_on(:nomis_staff_prisoner_check_enabled)
+        mock_service_with(StaffNomisChecker, checker)
+        expect(checker).to receive(:offender).and_return(offender)
+      end
```
Comment (StupidCodeFactory) [inline spec/decorators/visit_decorator_spec.rb]: I  think you can have  a cleaner test setup with only:

```ruby
let(:offender) { Nomis::Offender.new(id:1234)
before do
  switch_on(:nomis_staff_prisoner_check_enabled)
  mock_nomis_with(:lookup_active_offender, offender)
end
```
and remove the `StaffNomisChecker` double and the `expect` in the  `before` block

File: spec/models/visit_spec.rb [hunk 13]
```diff
@@ -343,4 +343,40 @@
       end
     end
   end
+
+  describe '#allowed_additional_visitors' do
+    let(:visitor1) { FactoryGirl.build_stubbed(:visitor) }
```
Comment (StupidCodeFactory) [inline spec/models/visit_spec.rb]: no need to add the `FactoryGirl` just use the short syntax, it looks a lot less cluttered and we don't have any other conflicting stubbing library.

File: spec/services/booking_responder/booking_request_processor_spec.rb [hunk 14]
```diff
@@ -13,17 +13,53 @@
 
   subject { described_class.new(staff_response) }
 
-  it '#process_request' do
-    expect {
-      subject.process_request message do
-        visit.accept!
+  describe '#process_request' do
+    context 'happy path' do
```
Comment (StupidCodeFactory) [inline spec/services/booking_responder/booking_request_processor_spec.rb]: happy path? Surely is a better way to describe the process. I have not idea what the "happy path" entails, same for the "sad path"

File: spec/services/booking_response_spec.rb [hunk 15]
```diff
@@ -0,0 +1,34 @@
+require 'rails_helper'
+
+RSpec.describe BookingResponse do
+  describe 'succesful' do
```
Comment (StupidCodeFactory) [inline spec/services/booking_response_spec.rb]: you're missing ".xxxxx" in front of all the methods descriptions

File: spec/services/nomis/api_spec.rb [hunk 16]
```diff
@@ -210,4 +210,48 @@
       expect(subject.first.id).to eq(first_contact.id)
     end
   end
+
+  describe 'book_visit' do
```
Comment (StupidCodeFactory) [inline spec/services/nomis/api_spec.rb]: `#book_visit`

File: spec/services/nomis/api_spec.rb [hunk 17]
```diff
@@ -210,4 +210,48 @@
...
+        override_restrictions: false,
+        client_unique_ref: 'visit_id_1234'
+      }
+    end
+
+    let(:offender_id) { 1_057_307 }
+
+    subject { super().book_visit(offender_id: offender_id, params: params) }
+
+    context 'happy path', vcr: { cassette_name: 'book_visit_happy_path' } do
```
Comment (StupidCodeFactory) [inline spec/services/nomis/api_spec.rb]:
```ruby
context 'With a context description of "sad path"' do
  it { expect(maintainer).to be_a_sad_panda }
end
```

File: spec/services/nomis/booker_spec.rb [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+  end
+
+  subject { described_class.new(visit) }
+
+  describe '#book' do
+    let(:booking_response) { subject.book }
+
+    describe 'api call to Nomis' do
+      it 'with the correct parameters' do
+        expect(Nomis::Api.instance).
```
Comment (StupidCodeFactory) [inline spec/services/nomis/booker_spec.rb]: why not extend the `mock_nomis_with` helper  method to accept an extra argument or block to perform argument call check?

File: spec/services/nomis/feature_spec.rb [hunk 19]
```diff
@@ -107,4 +107,32 @@
...
+    end
+
+    context 'when the book to nomis flag is enabled and the visit prison is not on the list' do
```
Comment (StupidCodeFactory) [inline spec/services/nomis/feature_spec.rb]: there need to be and outer context with the following description:

```ruby
describe 'the visit prison is not on the list' do
  context 'and when the book to nomis flag is enabled' 
  context 'and when the slot availability flag is enabled'
end
```

File: spec/services/nomis/feature_spec.rb [hunk 20]
```diff
@@ -107,4 +107,32 @@
...
+        switch_feature_flag_with(:staff_prisons_with_book_to_nomis, [])
+      end
+
+      it { expect(described_class.book_to_nomis_enabled?(prison_name)).to eq(false) }
+    end
+
+    context 'when the slot availability flag is enabled and the visit prison is not on the list' do
+      before do
+        switch_on :nomis_staff_book_to_nomis_enabled
+        switch_feature_flag_with(:staff_prisons_with_book_to_nomis, [prison_name])
```
Comment (StupidCodeFactory) [inline spec/services/nomis/feature_spec.rb]: I thought the prison wasn't on the list? Is this a bug?

File: app/controllers/concerns/staff_response_context.rb [hunk 21]
```diff
@@ -47,8 +48,9 @@ def visit_params
         :nomis_id,
         :banned,
         :not_on_list,
-        banned_until: %i[day month year]
-      ]
+        banned_until: [:day, :month, :year]
```
Comment (alan) [inline app/controllers/concerns/staff_response_context.rb]: Interesting, I think this must have been a rubocop autocorrection as I didn't change it. I'll revert and re-run rubocop to double check

File: app/controllers/prison/visits_controller.rb [hunk 22]
```diff
@@ -9,18 +9,26 @@ def process_visit
     @visit = load_visit.decorate
   end
 
+  # rubocop:disable Metrics/MethodLength
```
Comment (alan) [inline app/controllers/prison/visits_controller.rb]: Yes, it is ugly, suggestions welcome. This was something I wanted to get ideas on.

File: app/controllers/prison/visits_controller.rb [hunk 23]
```diff
@@ -9,18 +9,26 @@ def process_visit
...
   def update
     booking_response = booking_responder.respond!
+
     if booking_response.success?
-      flash[:notice] = t('process_thank_you', scope: %i[prison flash])
+      flash[:notice] = t('process_thank_you', scope: [:prison, :flash])
```
Comment (alan) [inline app/controllers/prison/visits_controller.rb]: @StupidCodeFactory here it has also changed from `%i` to explicit array. I'm pretty sure is rubocop

File: app/decorators/visit_decorator.rb [hunk 24]
```diff
@@ -45,6 +45,23 @@ def processed_at
     @processed_at ||= last_visit_state_change&.created_at || object.created_at
   end
 
+  def contact_list_working?
```
Comment (alan) [inline app/decorators/visit_decorator.rb]: No, this was wip as it dependend of yesterday's discussion. Coming soon with final logic

File: app/models/visit.rb [hunk 25]
```diff
@@ -1,6 +1,10 @@
 class Visit < ActiveRecord::Base
   extend FreshnessCalculations
   include PrincipalVisitor
+
+  attr_writer :book_to_nomis_opt_out
```
Comment (alan) [inline app/models/visit.rb]: We need to get the opt out logic to the booking responder acceptor class, the part that decides to make the api call or not, once is in that part of the code it only has access to the visit. Have you got any suggestions on how to improve this? I agree that is not ideal.

File: app/models/visit.rb [hunk 26]
```diff
@@ -1,6 +1,10 @@
...
+  attr_accessor :offender_id
```
Comment (alan) [inline app/models/visit.rb]: Good spot, I'll take it out

File: app/models/visit.rb [hunk 27]
```diff
@@ -84,6 +89,15 @@ class Visit < ActiveRecord::Base
     end
   end
 
+  def book_to_nomis_opt_out
+    # TODO: Changes in Rails 5 to `ActiveRecord::Type::Boolean.new.cast(string)`
+    ActiveRecord::Type::Boolean.new.type_cast_from_database(@book_to_nomis_opt_out)
```
Comment (alan) [inline app/models/visit.rb]: Yep, setter makes more sense.

Lets take a look together on other ways this could be done as you seem to have other ideas.

File: app/services/booking_responder/accept.rb [hunk 28]
```diff
@@ -5,8 +5,28 @@ def process_request(message = nil)
         visit.rejection = nil
         visit.accept!
 
-        BookingResponse.new(success: true)
+        if visit.book_to_nomis_opt_out?
```
Comment (alan) [inline app/services/booking_responder/accept.rb]: As discussed in person, the problem is not the POST failing, it is succeeding and then the pvb having an error. The idea with this approach is to delay as much as possible the API call to avoid  other errors leaving PVB and Nomis inconsistent.

File: app/services/booking_response.rb [hunk 29]
```diff
@@ -1,9 +1,44 @@
 class BookingResponse
-  def initialize(success:)
-    @success = success
+  ALREADY_PROCESSED_ERROR = 'already_processed'.freeze
+  PROCESS_REQUIRED_ERROR = 'process_required'.freeze
```
Comment (alan) [inline app/services/booking_response.rb]: I'll add a description on the intent of this class.

This class represents the result of the visit processing. Before the different states included the two above and were implicit. Now the response that staff can get includes extra states related to the result of the Nomis call (validation error / api error)

File: app/services/booking_response.rb [hunk 30]
```diff
@@ -1,9 +1,44 @@
...
+  NOMIS_VALIDATION_ERROR = 'nomis_validation_error'.freeze
+  NOMIS_API_ERROR = 'nomis_api_error'.freeze
+
+  attr_reader :error
+
+  def self.succesful
+    new(nil)
```
Comment (alan) [inline app/services/booking_response.rb]: Yes, this isn't clear I'll refactor it.

Regarding the pattern I think it's better wrapping the constants in class methods since it avoids getting unexpected values passed in and the actual values are an implementation detail

File: app/services/nomis/api.rb [hunk 31]
```diff
@@ -104,6 +104,20 @@ def fetch_contact_list(offender_id:)
       Nomis::ContactList.new(response)
     end
 
+    def book_visit(offender_id:, params:)
+      response = @pool.with { |client|
+        client.post(
+          "offenders/#{offender_id}/visits/booking", params, idempotent: false
```
Comment (alan) [inline app/services/nomis/api.rb]: Actually, I need to change this. Now the call is idempotent.

File: app/services/nomis/booker.rb [hunk 32]
```diff
@@ -0,0 +1,56 @@
+module Nomis
+  class Booker
```
Comment (alan) [inline app/services/nomis/booker.rb]: 👍 for the repository pattern

File: app/services/nomis/booker.rb (hunk 8 above)
Comment (alan) [inline app/services/nomis/booker.rb]: Yeah, I'll pass the arguments necessary for the call instead

File: app/services/nomis/booker.rb [hunk 33]
```diff
@@ -0,0 +1,56 @@
...
+    def call_api
+      self.booking = Nomis::Api.
+        instance.
+        book_visit(offender_id: offender_id, params: booking_params)
+    rescue Nomis::APIError
+      self.api_error = true
+    end
+
+    def build_booking_response
+      return BookingResponse.nomis_api_error if api_error
```
Comment (alan) [inline app/services/nomis/booker.rb]: This is related to the fact that we want to delay the API call as much as possible, so this class needs to return a `booking_response`

File: app/services/nomis/booking.rb [hunk 34]
```diff
@@ -0,0 +1,17 @@
...
+  class Booking
+    include NonPersistedModel
+
+    attribute :visit_id, Integer
+    attribute :error_message, String
+
+    def self.build(response)
+      if response.key?('error')
+
+        new(error_message: response['error']['message'])
```
Comment (alan) [inline app/services/nomis/booking.rb]: `ActiveModel::Error` is for validations so not a fan of adding outside the validation lifecycle. We use `.build` somewhere else to translate the api call response

File: app/services/nomis/client.rb [hunk 35]
```diff
@@ -27,9 +27,9 @@ def get(route, params = {})
       request(:get, route, params, idempotent: true)
     end
 
-  # def post(route, params)
-  #   request(:post, route, params)
-  # end
+    def post(route, params, idempotent:)
```
Comment (alan) [inline app/services/nomis/client.rb]: In our case is idempotent

File: app/services/nomis/client.rb [hunk 36]
```diff
@@ -80,17 +80,28 @@ def request(method, route, params, idempotent:)
     # rubocop:enable Metrics/MethodLength
     # rubocop:enable Metrics/AbcSize
 
+    def http_method_expects(method)
+      if method == :get
+        [200]
+      else
+        # TODO: Add 409 when idempotency key is ready
```
Comment (alan) [inline app/services/nomis/client.rb]: Reminder: Delete todos.

File: app/views/layouts/application.html.erb (hunk 9 above)
Comment (alan) [inline app/views/layouts/application.html.erb]: @chrishanes do you want to give it a go separating this into a different PR?

File: app/views/prison/visits/_book_to_nomis_opt_in.html.erb [hunk 37]
```diff
@@ -0,0 +1,27 @@
...
+    </label>
+  <% else %>
+    <% if book_to_nomis_config.already_booked_in_nomis? %>
+      <%= content_tag :p, t(".already_booked_in_nomis") %>
+    <% else %>
+      <%= content_tag :p, t(".book_to_nomis_not_possible") %>
+    <% end %>
+  <% end %>
+
+  <div id="rejection-message-opposite">
```
Comment (alan) [inline app/views/prison/visits/_book_to_nomis_opt_in.html.erb]: @chrishanes this does work as we expect.

Currently if you visit is set to be saved to Nomis when you select a time and then match the visitors these fields appear. Also using the checkbox does not trigger any changes.

The way I thought it was going to work was:
  - when the visit is approved, and the visit is set to be saved in nomis then staff wont see these fields.

  - when the visit is approved, and staff opt out from saving to nomis, or the visit can't be saved to nomis then staff will see these 2 fields

  - prisons that don't have book to nomis feature enabled will see no changes

File: config/locales/en/prison_views.yml [hunk 38]
```diff
@@ -25,9 +26,28 @@ en:
...
+          This visit wasn't booked.
+        </h1>
+        <p>Please either reject or accept the booking</p>
       at_least_one_estate: You must select at least one estate
       invalid_estate_selection: You don't access to these estates
+      nomis_validation_error_html: |
+        <h1 class="heading-medium error-summary-heading" id="error-summary-heading">
+          This visit wasn't booked.
+        </h1>
+        <p>Choose a rejection reason to reject the visit or <a href="#nomis-opt-out">book this visit anyway</a>.</p>
```
Comment (alan) [inline config/locales/en/prison_views.yml]: @kayesoutham @kyliehavelock I'm wondering if we should tweak this validation error message. If the prisoner has a restriction, such as a Closed restriction, staff won't really know what validation failed. I think it would be helpful to explain that this validation error can happen when the offender has restrictions and to check in Nomis. What do you think?

**Comment by alan** [comment]
@StupidCodeFactory can you review again please?

File: app/services/booking_responder.rb [hunk 39]
```diff
@@ -57,4 +58,17 @@ def bookable?
...
       visit.slot_granted?
   end
+
+  def persist_to_nomis=(val)
+    # TODO: Changes in Rails 5 to `ActiveRecord::Type::Boolean.new.cast(string)`
+    @persist_to_nomis = ActiveRecord::Type::Boolean.new.type_cast_from_database(val)
+  end
+
+  def persist_to_nomis?
+    @persist_to_nomis
```
Comment (StupidCodeFactory) [inline app/services/booking_responder.rb]: this can return nil `!!@persist_to_nomis`

File: app/views/prison/visits/_book_to_nomis_opt_in.html.erb [hunk 40]
```diff
@@ -0,0 +1,29 @@
+<div class="panel" id="nomis-opt-out">
+  <% if book_to_nomis_config.possible_to_book? %>
+    <%= content_tag :p, t(".book_to_nomis_opted_in_by_default") %>
```
Comment (StupidCodeFactory) [inline app/views/prison/visits/_book_to_nomis_opt_in.html.erb]:
```erb
<p>
  <%=  t(".book_to_nomis_opted_in_by_default") %>
</p>
```

File: app/views/prison/visits/_book_to_nomis_opt_in.html.erb [hunk 41]
```diff
@@ -0,0 +1,29 @@
...
+    <%= hidden_field_tag :book_to_nomis_opted_in, true %>
+    <fieldset>
+      <div class="multiple-choice push-bottom">
+        <%= check_box_tag :book_to_nomis_opted_in, false, !book_to_nomis_config.opted_in?, class: 'js-BookToNomis', 'data-el': 'booktonomis-message', id: 'book_to_nomis_opt_in' %>
+        <label for="book_to_nomis_opt_in"><%= t('.dont_book_to_nomis') %></label>
+      </div>
+    </fieldset>
+  <% else %>
+    <% if book_to_nomis_config.already_booked_in_nomis? %>
+      <%= content_tag :p, t(".already_booked_in_nomis") %>
```
Comment (StupidCodeFactory) [inline app/views/prison/visits/_book_to_nomis_opt_in.html.erb]: same as above

File: app/models/staff_response.rb [hunk 42]
```diff
@@ -22,15 +22,16 @@ class StaffResponse
...
+      except: [
+        :created_at,
+        :updated_at,
+        :slot_granted,
+        :slot_option_0,
+        :slot_option_1,
+        :slot_option_2,
+        :nomis_id,
+        :human_id
+      ],
```
Comment (alan) [inline app/models/staff_response.rb]: Don't think so, rubocop is passing
